### PR TITLE
[decoupled-execution] simplify code

### PR DESCRIPTION
**Pull request by zekun000**
1. Introduce find_elem to simplify a few places that requires traversing the list to find matching element.
2. Change sending request from eagerly to lazily.
3. A few cleanup

## Motivation

(Write your motivation for proposed changes here.)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

## Test Plan

(Share your test plan here. If you changed code, please provide us with clear instructions for verifying that your changes work.)

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/diem/diem/tree/main/developers.diem.com, and link to your PR here.)

## If targeting a release branch, please fill the below out as well

 * Justification and breaking nature (who does it affect? validators, full nodes, tooling, operators, AOS, etc.)
 * Comprehensive test results that demonstrate the fix working and not breaking existing workflows.
 * Why we must have it for V1 launch.
 * What workarounds and alternative we have if we do not push the PR.

### CODE REVIEW

File: consensus/src/experimental/buffer_manager.rs [hunk 1]
```diff
@@ -172,38 +174,19 @@ impl StateManager {
...
-                if !matches!(&*buffer_item, BufferItem::Executed(_)) {
-                    break;
-                }
-            }
-            cursor = get_next(&cursor);
-        }
-        self.execution_root = cursor;
+    fn advance_executed_root(&mut self) {
+        let cursor = self.execution_root.clone();
+        self.execution_root = find_elem(cursor, |item| !matches!(*item, BufferItem::Executed(_)));
```
Comment (yuxiamit) [inline consensus/src/experimental/buffer_manager.rs:179]: We probably also need to move the root if the current buffer item is signed. Recall that we allow scattered signed buffer items appear after the signing root:

[Signed] -- [Signed] -- [Executed: Signing Root] -- [Executed] -- [Signed] -- [Executed] -- [Ordered] ...

Now the executed root will stop at the 5-th but it should be at 7-th.

File: consensus/src/experimental/buffer_manager.rs [hunk 2]
```diff
@@ -172,38 +174,19 @@ impl StateManager {
...
-                    break;
-                }
-            }
-            cursor = get_next(&cursor);
-        }
-        self.signing_root = cursor;
+    fn advance_signing_root(&mut self) {
+        let cursor = self.signing_root.clone();
+        self.signing_root = find_elem(cursor, |item| {
+            !matches!(*item, BufferItem::Signed(_) | BufferItem::Aggregated(_))
```
Comment (yuxiamit) [inline consensus/src/experimental/buffer_manager.rs:188]: Right now there should not be an aggregated item left on the buffer. We dequeue them right away. Maybe we need to change this to Signed only.

File: consensus/src/experimental/buffer_manager.rs [hunk 3]
```diff
@@ -284,30 +267,13 @@ impl StateManager {
         &mut self,
         executed_blocks: Vec<ExecutedBlock>,
     ) -> anyhow::Result<()> {
-        if self.execution_root.is_none() {
-            // right after a sync
-            return Ok(());
-        }
-
-        let current_cursor = self.execution_root.clone();
+        let current_cursor = find_elem(self.execution_root.clone(), |item| {
```
Comment (yuxiamit) [inline consensus/src/experimental/buffer_manager.rs:270]: Shall we keep the comments on why we are comparing the last id?

File: consensus/src/experimental/linkedlist.rs [hunk 4]
```diff
@@ -207,6 +207,17 @@ pub fn set_elem<T>(link: &Link<T>, new_val: T) {
     node.elem.replace(new_val);
 }
 
+pub fn find_elem<F: Fn(&T) -> bool, T>(link: Link<T>, compare: F) -> Link<T> {
+    let mut current = link.clone();
```
Comment (yuxiamit) [inline consensus/src/experimental/linkedlist.rs:211]: Do we need clone here? the function owns `link`?

File: consensus/src/experimental/buffer_manager.rs (hunk 2 above)
Comment (zekun000) [reply consensus/src/experimental/buffer_manager.rs:188]: yeah I realized this too

**Comment by zekun000** [comment]
@yuxiamit thanks for reviewing, will update. I'm also doing another bigger change on top of this, will send out later which basically bundling the root moving with sending requests to pipeline i.e. move execution root from A -> B would trigger a execution request for B, move signing root from A -> B would trigger a signing request.

File: consensus/src/experimental/buffer_manager.rs [hunk 5]
```diff
@@ -202,12 +206,14 @@ impl StateManager {
...
-                    callback: aggregated.callback,
-                }))
-                .ok();
+                self.persisting_phase_tx
+                    .send(PersistingRequest {
+                        blocks,
+                        commit_ledger_info: aggregated.aggregated_proof,
+                        callback: aggregated.callback,
+                    })
+                    .await
```
Comment (yuxiamit) [inline consensus/src/experimental/buffer_manager.rs:215]: This future await is dangerous that we need make sure no other async task will visit this item before it yields. Right now, this function is directly called from the main loop, so it is okay.

File: consensus/src/experimental/buffer_manager.rs [hunk 6]
```diff
@@ -500,61 +377,50 @@ impl StateManager {
...
+                    self.advance_execution_root().await;
+                    if self.signing_root.is_none() {
+                        self.advance_signing_root().await;
+                    }
                 }
-                Some(signing_resp) = self.signing_phase_rx.next() => {
-                    self.process_signing_response(signing_resp).await
+                Some(response) = self.signing_phase_rx.next() => {
+                    self.process_signing_response(response).await;
+                    self.advance_signing_root().await;
```
Comment (yuxiamit) [inline consensus/src/experimental/buffer_manager.rs:413]: shall we try advancing only when the signing succeeded?

File: consensus/src/experimental/buffer_manager.rs [hunk 7]
```diff
@@ -500,61 +377,50 @@ impl StateManager {
...
                 }
                 Some(commit_msg) = self.commit_msg_rx.next() => {
-                    self.process_commit_msg(commit_msg).await
+                    self.process_commit_message(commit_msg).await;
+                    self.try_advance_head().await;
```
Comment (yuxiamit) [inline consensus/src/experimental/buffer_manager.rs:417]: similarly, advance the head only if the commit message matched an item?

File: consensus/src/experimental/buffer_manager.rs (hunk 5 above)
Comment (zekun000) [reply consensus/src/experimental/buffer_manager.rs:215]: we should move all channels to sync in following pr

File: consensus/src/experimental/buffer_manager.rs (hunk 6 above)
Comment (zekun000) [reply consensus/src/experimental/buffer_manager.rs:413]: this would retry if the signing failed

**Review by yuxiamit — APPROVED** [review]

File: consensus/src/experimental/buffer_manager.rs (hunk 7 above)
Comment (zekun000) [reply consensus/src/experimental/buffer_manager.rs:417]: don't have strong opinion on this, we can return something if commit message aggregated an item and use advance_head

File: consensus/src/experimental/buffer_manager.rs [hunk 8]
```diff
@@ -137,130 +137,104 @@ impl StateManager {
...
+                        .expect("Failed to send signing request");
                 }
+                _ => unreachable!(),
             }
-            cursor = get_next(&cursor);
         }
-        self.signing_root = cursor;
     }
 
     /// Pop the prefix of buffer items until (including) aggregated_item_cursor
```
Comment (yuxiamit) [inline consensus/src/experimental/buffer_manager.rs:195]: aggregated_cursor

File: consensus/src/experimental/buffer_manager.rs [hunk 9]
```diff
@@ -137,130 +137,104 @@ impl StateManager {
...
-    /// Prepare an aggregated item and send it to the persisting phase
-    fn try_persisting_blocks(&mut self, aggregated_item_cursor: BufferItemRootType) {
-        let mut cursor = self.buffer.head.as_ref().cloned();
+    /// Send persist request.
+    fn advance_head(&mut self, aggregated_cursor: BufferItemRootType) {
```
Comment (yuxiamit) [inline consensus/src/experimental/buffer_manager.rs:197]: Maybe we simply pass a block id here? We got the element right in the `process_commit_message `, and here we destruct it again just for the id.

File: consensus/src/experimental/buffer_manager.rs [hunk 10]
```diff
@@ -137,130 +137,104 @@ impl StateManager {
...
-                }))
-                .ok();
-            } else {
-                panic!("Bad Aggregated Item Cursor: Item is not aggregated.");
+            if item.block_id() == target_block_id {
+                if let BufferItem::Aggregated(aggregated) = item {
+                    block_on(self.persisting_phase_tx.send(PersistingRequest {
+                        blocks: blocks_to_persist,
+                        commit_ledger_info: aggregated.aggregated_proof,
+                        // we use the last callback
```
Comment (yuxiamit) [inline consensus/src/experimental/buffer_manager.rs:217]: I'll add a TODO to add some comments at the BlockStore side that the callback needs to be okay if only the last one gets called.

File: consensus/src/experimental/buffer_manager.rs [hunk 11]
```diff
@@ -308,268 +279,112 @@ impl StateManager {
...
-                            }
-                            break;
+            VerifiedEvent::CommitVote(vote) => {
+                // find the corresponding item
+                let current_cursor = find_elem(self.buffer.head.clone(), |item| {
+                    item.block_id() == vote.commit_info().id()
+                });
+                if current_cursor.is_some() {
+                    let mut buffer_item = take_elem(&current_cursor);
+                    let new_item = match buffer_item.add_signature_if_matched(
```
Comment (yuxiamit) [inline consensus/src/experimental/buffer_manager.rs:353]: looks like we can directly pass a `CommitVote`

**Review by yuxiamit — DISMISSED** [review]

**Review by yuxiamit — APPROVED** [review]

**Review by JoshLind — APPROVED** [review]

**Pull request closed without merging**